src/components/filters/AttributeFilter/AttributeElements.tsx

```diff
@@ -0,0 +1,226 @@
+        if (event) {
+            event.preventDefault();
+        }
+        // do not execute while still loading
+        if (this.state.isLoading) {
+            return;
+        }
+        this.setState({
+            isLoading: true
+        });
+        const offset = get(this.state, 'validElements.paging.offset', 0);
+        const count = get(this.state, 'validElements.paging.count', 0);
+        const nextOffset = offset + count;
+        this.getValidElements(this.props, nextOffset);
+    }
+
+    public getValidElements(props: IAttributeElementsProps, offset: number) { // IAttributeElementsProps
+        const { projectId, options, uri } = props;
+        const optionsWithUpdatedPaging = {
+            ...options,
+            offset
+        };
+
+        const identifierPromise = new Promise((resolve, reject) => {
+            return (props.identifier || this.identifier)
+                ? resolve(props.identifier || this.identifier)
+                : props.metadata.getIdentifiersFromUris(projectId, [uri])
+                .then(
+                    (result) => {
+                        this.identifier = result[0].identifier;
+                        resolve(this.identifier);
+                    },
+                    (error) => {
+                        reject(error);
+                    }
+                );
+        });
+
+        identifierPromise
+            .then((identifier: string) => props.metadata.getValidElements(
+                projectId,
+                identifier,
+                optionsWithUpdatedPaging
+            ))
+            .then((response: GoodData.IValidElementsResponse) => {
+                const items = [
+                    ...get(this.state, 'validElements.items', []),
+                    ...response.validElements.items
+                ];
+                const offset = get(
+                    this.state,
+                    'validElements.paging.offset',
+                    parseInt(response.validElements.paging.offset, 10) || 0
+                );
+                const mergedResponse = {
+                    ...response.validElements,
+                    items,
+                    paging: {
+                        total: parseInt(response.validElements.paging.total, 10),
+                        offset,
+                        count: items.length
+                    }
+                };
+                this.setState({
+                    validElements: mergedResponse,
+                    isLoading: false
+                });
+            }, error => (
+                this.setState({
+                    error,
+                    isLoading: false
+                })
+            ));
+    }
+
+    public render() {
+        const { validElements, isLoading, error } = this.state;
+        return this.props.children({
+            validElements,
+            loadMore: this.loadMore,
+            isLoading,
+            error
+        });
```

Comment:
Why object? I would prefer simple parameters. This adds unnecessary complexity and also i cant see any usecase (ie skiping one of the params in consumer / too many of them).